Develop

wch-mirek opened the pull request:
Made changes accordingly to feedback from [Redbooth task](https://redbooth.com/a/#!/projects/2135400/tasks/48960839)

wch-mirek committed:
eslint added and adjusted code to those rules

wch-mirek committed:
Missing Paw types definitions

wch-mirek committed:
Bearer auth and branch renaming to "develop"

wch-mirek committed:
OAuth2 handler

wch-mirek committed:
Other body types converter

wch-mirek committed:
Adjustments to Micha feedback

wch-mirek committed:
Param name typo fix

wch-mirek committed:
Fixed some eslint errors

wch-mirek committed:
Fixed some eslint errors
Added default values for OAuth2 authorization

mittsh commented on src/lib/paw-to-openapi-converter/components/parameters-converter.ts:
```diff
@@ -1,122 +1,125 @@
-          in: "header",
-          schema
-        })
+          in: 'header',
```
This part is extremely buggy…. 😔

With this set up:

![image](https://user-images.githubusercontent.com/1319906/91656192-791fc600-eabf-11ea-8616-9b5332c6c5cd.png)

![image](https://user-images.githubusercontent.com/1319906/91656195-7f15a700-eabf-11ea-8537-b47d7895bed5.png)

![image](https://user-images.githubusercontent.com/1319906/91656200-863cb500-eabf-11ea-812f-ff72a23ced6a.png)

![image](https://user-images.githubusercontent.com/1319906/91656206-981e5800-eabf-11ea-8644-a6efd51ad2ef.png)

This is the result I'm getting:

```json
{
  "openapi": "3.0.3",
  "info": {
    "title": "OpenAPI Generator.paw",
    "version": "1598781186333"
  },
  "paths": {
    "/api/v3/test/{My-Header}/{query}/": {
      "servers": [
        {
          "url": "https://example.com/"
        }
      ],
      "post": {
        "operationId": "5f0156e0-a25a-43d9-a75f-21f627509df9",
        "summary": "Some Request",
        "description": "Some description",
        "responses": {
          "default": {
            "description": "Default response"
          }
        },
        "parameters": [
          {
            "name": "query",
            "in": "query",
            "schema": {
              "type": "string",
              "default": "QueryParamValue"
            }
          },
          {
            "name": "My-Header",
            "in": "path",
            "required": true,
            "schema": {
              "type": "string",
              "default": "header-value"
            }
          },
          {
            "name": "query",
            "in": "path",
            "required": true,
            "schema": {
              "type": "string",
              "default": "QueryParamValue"
            }
          },
          {
            "name": "My-Header",
            "in": "header",
            "schema": {
              "type": "string",
              "default": "header-value"
            }
          },
          {
            "name": "Content-Type",
            "in": "header",
            "schema": {
              "type": "string",
              "default": "application/json; charset=utf-8"
            }
          }
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "example": {
                "value": {
                  "test": "value"
                }
              }
            }
          }
        }
      }
    }
  },
  "components": {
    "securitySchemes": {},
    "examples": {}
  }
}
```

So apparently, the whole way parameters are handled isn't clear and produces erroneous results:

1. Params that have nothing to do with `path` params end up in the URL (`/api/v3/test/{My-Header}/{query}/`)

2. The same param `query` (value = `QueryParamValue`) appears twice, once as `query` (which is correct) and once as `path` (incorrect)

3. The same happens for param `My-Header` which appears in `header` (correct) and `path` (incorrect)

4. Descriptions are missing when present in Paw

Though, path variables seem to be working fine:

![image](https://user-images.githubusercontent.com/1319906/91656285-488c5c00-eac0-11ea-89f1-afeda3221ed2.png)

wch-mirek replied:
Let me provide conclusions from my debugging.

_Test case:_
<img width="515" alt="+ POST https example com Path variable" src="https://user-images.githubusercontent.com/69789351/92101912-1c355000-edde-11ea-95e8-cfac4a0d15f9.png">
<img width="518" alt="POST httpsexample comPath variable" src="https://user-images.githubusercontent.com/69789351/92101931-222b3100-edde-11ea-942f-1940b3acfad6.png">
<img width="509" alt="POST httpsexample comPath variable" src="https://user-images.githubusercontent.com/69789351/92101938-248d8b00-edde-11ea-95e5-25a2bbf21239.png">

_Problems:_
I haven’t noticed that headers and query parameters are also stored in „request.variables” as seen on first screenshot. They goes there after adding descriptions which I haven’t done and that is why I missed that issues during development. I thought that „request.variables” only stores path variables and that is why duplicates of headers and query params were added in shown test case.


**How to solve?**


_What data I have during generating output:_
1. request.urlParameters / request.urlParameterNames - which stores values/names of query params
2. request.headers - which stores key<->value map of headers
3. request.variables - which stores information about all - query params, headers and path variables

_What I need to actually output is:_
Put urlParameters, headers and path variables into OpenApi’s operations.parameters with correct „in” param that describes which one is it.

_What problem needs to be solved:_
Path variables do not have separate place in Paw’s request or do not have any „type” or „in” param to be able to distinguish them from other objects (headers, query params) - the request.variables type is defined [here](https://github.com/luckymarmot/Paw-OpenAPI3Generator/blob/74ca00f7904e6631e688bfeaafb2c7265e6181a3/src/types-paw-api/paw.d.ts#L336)

_Propose of solution:_
1. First copy request.variables to local variable (let's name it $requestVariables)
2. Then parse headers and query params  using request.urlParameter(Names) and request.headers
3. While parsing headers and query params try to fetch attached variable to each one For example using [code snippet](https://github.com/luckymarmot/Paw-PostmanCollectionV2Generator/blob/8a153f526ddd37717eaf0d8fe819f014ad866cd0/src/lib/convertHeaders.ts#L11) from Postman Generator (this snippet is for headers).
4. If variable for header or query param will be found, remove it from $requestVariables
5. Parse all $requestVariables that are still left and treat them as path variables.

@mittsh Please take a look at that note and confirm if that way of solving this problem seems ok for You or maybe if You have any additional concerns about that.

wch-mirek replied:
Ok I got another case that is related to this problem:
![image](https://user-images.githubusercontent.com/69789351/92108979-8226d500-ede8-11ea-998e-21994684224e.png)

_Problems:_
1. The URL in PawRequest has already replaced values, so e.g request.urlBase returns in above example `http://developers.uspto.gov/` instead of `{scheme}://{subdomain}upsto.gov/`. It can make it impossible to distinguish if both of those variables will have same value e.g. 'http'.
2. That variables are also stores in request.variables, but that seems fine since they should be treated same as path variables in example from previous comment.


**How to solve?**


_What data I have during output:_
1. request.urlBase which contain already replaced variables
2. $requestVariables (from previous comment) that should store all path variables left to be processed

_What problems needs to be solved:_
1. While parsing path parameters they needs to be additionally splitted:
a. path params that are in hostname part of URL should be passed to OpenAPI server variables
b. path params that are outside hostname pare of URL should be passed to OpenAPI request variables

_Propose of solution:_
Unfortunatelly for now I have no proposal for that because of that URL already have replaced variables which complicates that situation. For now I'm working on importer [related case](https://github.com/luckymarmot/Paw-OpenAPI3Importer/pull/1#discussion_r478450156) and maybe I'll come up with some solution while working there

mittsh commented on src/lib/paw-to-openapi-converter/components/auth-converter.ts:
```diff
@@ -1,110 +1,173 @@
+      const securityExample: OpenAPI.ExampleObject = {
+        summary: OAuth2CredentialsLabel,
+        value: restOauth2Values,
+      };
```
Thinking about this once again, I wouldn't export this security example and the `restOauth2Values`. The reason being that people exporting from Paw to OpenAPI 3.0 are probably willing to share the document with colleagues or 3rd parties. Better not share here credentials (token, refresh token) which anyways should be easily recreated using the Authorization URL and Token URL.

If that was an export to Postman or cURL, I would see this differently: people exporting from one API client to another want to be able to send the exact same requests. But when generating an OpenAPI description file, they want to _describe_ the API and not _send request_.

mittsh commented on src/lib/paw-to-openapi-converter/components/auth-converter.ts:
```diff
@@ -1,110 +1,173 @@
+      } = oauth2;
+
+      this.requirement = {};
+      this.key = 'OAuth2';
```
This is quite incorrect. If one has two _**very different**_ authentication methods using OAuth 2 (or Basic Auth for that matter), they are mixed up in the output `securitySchemes`.
This is an important use case. At the same time, we don't have to create X entries in `securitySchemes`, if they are all the same.

A suggestion would be to:
1. Create a simple hash of the security scheme (for example, can be first few characters of the md5 of `authorization_uri + authorization_uri + scope`)
2. Use this hash as key in addition to OAuth 2, you could name the key `OAuth2_{hash}`
3. A bonus would be: right before finishing the import process, to check how many auths of one kind do we have. If only one `OAuth2`, then rename it from `OAuth2_{hash}` to simply `OAuth2` to keep a nice output

Open to any other implementation suggestions, as long as we keep all the possible `securitySchemes` if they end up being different.

wch-mirek replied:
@mittsh I would just want to discuss one thing here - OpenAPI documentation page have this authorization form view looking like here:
<img width="662" alt="image" src="https://user-images.githubusercontent.com/69789351/91987864-e08f7d00-ed2e-11ea-91a9-3efb35f2ae08.png">

That "key" value is used here as form label, so using "OAuth2 {hash}" might look strange for user. Maybe we could use something more human-readable like "OAuth2 {request_name}"? What do You think about it @mittsh ?

mittsh replied:
@wch-mirek That's a good point! Though, I'm afraid that if a user has the exact same auth mechanism in 20 different requests, we'll create 20 different entries for OAuth 2 when we could have a single one. That's where the idea of the hash came from. Using a hash makes sure that:
1. Two different auths will never be mixed
2. If many requests use the same auth, we don't recreate auth configs needlessly

Perhaps you could use a hash internally to have this behavior but just use `{request_name}` as a label for the user (picking the 1st request that used this auth).

This solution would work for me if you like it as well 😄

wch-mirek replied:
@mittsh Hm ok I think that tool is mostly for developers, so hashes in labels shouldn't be that scary and also I tried to come up with solution of putting 1st request name as authorization label and instead of hashing values - simple compare values. But I stumble on big if like so:

```
if (
  !sameAuthFound
  && oauth2?.grant_type === AuthConverter.capitalToCamelCase(grantType)
  && oauth2?.authorization_uri === (flow as OpenAPI.OAuthFlowObject).authorizationUrl
  && oauth2?.access_token_uri === (flow as OpenAPI.OAuthFlowObject).tokenUrl
  && oauth2.scope === AuthConverter.convertOpenAPIToPawScopes(
    (flow as OpenAPI.OAuthFlowObject).scopes,
  )
)
```

Using hashes will be more elegant and it will shorten logic to comparing just labels for authorizations. Then I'll go here with 'OAuth2 {hash}' solution then :)

mittsh replied:
Good 👍

mittsh reviewed (commented):
Please see the comments below 😊

wch-mirek committed:
Generating using EnvironmentManager

wch-mirek committed:
For now static URL if it is based fully on dynamic value - URL regex do not work with variables

pongstr committed:
init

pongstr committed:
init: update

pongstr committed:
init: update setup

pongstr committed:
chore(update): simplify converters

pongstr committed:
init: update build configs

pongstr committed:
init: update utils to match current flow

pongstr committed:
init: temp add param converters

pongstr committed:
init: safe-parse json string

pongstr committed:
init: fix for oas3 validation

pongstr committed:
updates

pongstr committed:
init: security schema (oauth2)

- updated: nullable types
- updated: safe parse json string or query string

pongstr committed:
updates

pongstr committed:
init: refactor param converter

pongstr committed:
chore(refactor):  init

- updated: improve code coherence
- updated: improve webpack+typescript setup, eslint
- updated: convert paw request to a valid openapi v3 document
- added: generate json schema based on request info (body, parameters, query)
- added: generate openapi response definition based on last httpExchange data
- refactor: parameter converters, moved to converter file instead

### TODO:

- [ ] [SecurityRequirementObject](https://swagger.io/specification/#security-scheme-object) type: `apiKey`, `http`
-  `apiKey` is not available in `Paw.Request` which is a key to evaluate whether 
items in a request parameters are security key/value.
- `http` is not available in `Paw.Request` as well, however, is it safe to assume `Authorization` header key
is a "Bearer" scheme?
